memtagabielf64/memtagabielf64.rst

```diff
@@ -0,0 +1,723 @@
+  +----------------------------------------+----------------+
+
+``SHT_AARCH64_MEMTAG_GLOBALS_STATIC`` is a section type used during static
+linking. This section is of size ``0``. For each tagged global variable in the
+object file, there exists an ``R_NONE`` relocation that references both the
+symbol to be tagged, and this section.
+
+``SHT_AARCH64_MEMTAG_GLOBALS_DYNAMIC`` is a section type used during dynamic
+loading. For each linked object, there should only be a single
+``SHT_AARCH64_MEMTAG_GLOBALS_DYNAMIC`` section, which is pointed to by the
+``DT_AARCH64_MEMTAG_GLOBALS`` dynamic array entry. This section contains a
+compressed series of "global variable descriptors", metadata encoding the offset
+and size to which a random memory tag should be applied. This section is defined
+in detail later in this document.
+
+The value is in the AArch64 Processor-specific range. The value is
+subject to change if there is a clash with any section types added by
+AAELF64_.
+
+Tagging Global Variables
+========================
+
+Compilation
+-----------
+
+Global variables that wish to be tagged must meet MemtagABI-specific guidelines.
+In particular, a compiler that destines a global variable to be tagged at
+runtime must:
+
+1. If necessary, round up the size such that it's a multiple of the tag granule
+   size (i.e. ``size % 16 == 0``), and
+
+2. If necessary, round up the alignment such that it's a multiple of the tag
+   granule size (i.e. ``alignment % 16 == 0``).
+
+Static Linking
+--------------
+
+If the translation unit contains tagged global variables, the compiler should
+produce an ``SHT_AARCH64_MEMTAG_GLOBALS_STATIC`` section, as well as a
+relocation section with ``sh_link`` pointing to the
+``SHT_AARCH64_MEMTAG_GLOBALS_STATIC`` section.
+
+For each tagged global variable, the compiler should produce an ``R_NONE``
+relocation placed in the aforementioned relocation section. This relocation
+should have its ``ELF64_R_SYM`` bits of the ``r_info`` field point to the global
+variable that needs to be tagged. This communicates to the linker that the
+global variable in question is marked for tagging.
+
+In addition, the linker should ensure that all references to tagged global
+variable are via. the GOT (including for internal symbols). The relocations
+described later in this document ensure that the address tag for a global
+variable is materialized into the GOT, resulting in correctly tagged global
+variable accesses.
+
+Dynamic Linking
+---------------
+
+The dynamic linker should parse the ``SHT_AARCH64_MEMTAG_GLOBALS_STATIC``
+section to identify all tagged global variables.
+
+Resolving Multiple Definitions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Ideally, all translation units linked into a program are all built with or
+without tagged globals. This may not always be possible however, and so the
+linker must take special care when resolving the definition of a global variable
```

Comment:
Sorry for the confusion, this section describes symbol resolution for multiple mismatching declarations, not multiple mismatching definitions.

If the definition is tagged, and one of the uses is untagged (i.e. the declaration is also untagged), then things can go wrong. One example is that a tagged symbol is generated, but there is a reference in hand-written assembly that's untagged, and the assembly uses a PC-relative reference rather than a GOT-relative reference to the global. An example from HWASan (a software implementation of MTE) is here: https://boringssl-review.googlesource.com/c/boringssl/+/36924

Hence, when we have mismatched declarations, we do the conservative thing and demote the global to be untagged. Updated the text to try and be more clear, please LMK.

Reply:
Ack. Thanks for the update. It is clear that there are declarations involved and not multiple definitions.

A related thought to an earlier comment about a different approach using a set of GOT generating relocations. We'd need something like :tag: and :tag_lo12: that an assembly file would need to use these to get the tagging behaviour.

Reply:
Making the linker inspect all relocatable files to check a symbol is an unusual request to the linker. For efficient linker design, it just discards all non-prevailing COMDAT groups. Is this really the responsibility of a linker?

Perhaps I have missed a big picture. A companion design doc about how this is imagined to be used in the llvm toolchain with a concrete example (test source code, linker command line, runtime behavior, etc) may help. Some components may not be the scope of this ABI doc, though.

Reply:
The intention is not to scan symbols inside of discarded COMDAT groups - it's just to keep track of the tagged-ness of the declarations and use that information when selecting the definition. This is seemingly pretty cheap in LLD, a small update to `Symbol::mergeProperties`. I'll upload the tentative LLD patches soon so you can take a look.